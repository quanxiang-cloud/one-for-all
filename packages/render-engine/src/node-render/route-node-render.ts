import React, { useContext } from 'react';

import NodeRender from '.';
import RouteContext from './route-context';
import { useLifecycleHook } from './hooks';
import type { CTX, RouteNode, SchemaNode } from '../types';
import HTMLNodeRender from './html-node-render';
import { render } from '@testing-library/react';

export interface Props {
  node: RouteNode;
  ctx: CTX;
  parentRoute?: string;
}

function RouteNodeRender({ node, ctx, parentRoute }: Props): React.ReactElement | null {
  useLifecycleHook(node.lifecycleHooks || {});
  const currentRoutePath = `${parentRoute}/${node.path}`.replace(/\/+$/, '').replace(/^\/*/, '/');

  console.log(ctx.routeState.location.pathname, currentRoutePath, parentRoute);
  
  if (ctx.routeState.location.pathname === currentRoutePath) {
    return React.createElement(NodeRender, { node: node.node, ctx, parentRoute: currentRoutePath, });
  }

  // if (node.node.type !== 'route-node') {
  //   return React.createElement(NodeRender, { node: node.node, ctx, parentRoute: `${parentRoute}/${node.path}`, });
  // }

  // if (parentRouteNodePath === node.path) {
  //   if (node.node.type === 'html-element') {
  //     return React.createElement(NodeRender, { node: node.node, ctx });
  //   }
  // }

  // const match = matches.find(({ path }) => {
  //   return path.replace(/\/+$/, '').replace(/^\/*/, '/') === ctx.routeState?.location.pathname;
  // });

  // console.log(ctx.routeState);
  // if (!match) {
  //   return null;
  // }

  // return React.createElement(NodeRender, { node: match.element as SchemaNode, ctx });
  return null;
}

export default RouteNodeRender;
