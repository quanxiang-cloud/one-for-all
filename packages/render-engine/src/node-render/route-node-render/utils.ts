// trim begin and last slash
export function trimSlash(path: string): string {
  return path.replace(/^\/|\/$/g, '');
}

export function isParamHolder(fragment: string): boolean {
  return /^:[a-zA-Z_][[a-zA-Z_$0-9]+$/.test(fragment);
}
