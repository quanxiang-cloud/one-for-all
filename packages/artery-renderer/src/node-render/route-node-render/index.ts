import React, { useContext } from 'react';

import NodeRender from '../index';
import RoutePathContext from './route-path-context';
import useMatch from './use-match';
import { trimSlash } from './utils';
import { useLifecycleHook } from '../hooks';
import type { RouteNode } from '../../types';
import useCTX from '../../use-ctx';

export interface Props {
  node: RouteNode;
}

function buildCurrentPath(parentPath: string, routePath: string): string {
  if (parentPath === '/') {
    return `/${trimSlash(routePath)}`;
  }

  return `${parentPath}/${trimSlash(routePath)}`;
}

function RouteNodeRender({ node }: Props): React.ReactElement | null {
  const ctx = useCTX();
  useLifecycleHook(node.lifecycleHooks || {});

  const parentRoutePath = useContext(RoutePathContext);
  const currentRoutePath = buildCurrentPath(parentRoutePath, node.path);
  const match = useMatch(ctx.location$, currentRoutePath, node.exactly ?? false);

  if (match) {
    return React.createElement(
      RoutePathContext.Provider,
      { value: currentRoutePath },
      React.createElement(NodeRender, { node: node.node, ctx }),
    );
  }

  return null;
}

export default RouteNodeRender;
