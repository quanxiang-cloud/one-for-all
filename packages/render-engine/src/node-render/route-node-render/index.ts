import React, { useContext } from 'react';

import NodeRender from '../index';
import RoutePathContext from './route-path-context';
import useMatch from './use-match';
import { trimSlash } from './utils';
import { useLifecycleHook } from '../hooks';
import type { CTX, RouteNode } from '../../types';

export interface Props {
  node: RouteNode;
  ctx: CTX;
}

function RouteNodeRender({ node, ctx }: Props): React.ReactElement | null {
  useLifecycleHook(node.lifecycleHooks || {});

  const parentRoutePath = useContext(RoutePathContext);
  const currentRoutePath = `${parentRoutePath}/${trimSlash(node.path)}`;
  const match = useMatch(ctx.historyState.location$, currentRoutePath, node.exactly ?? false);

  if (match) {
    return React.createElement(
      RoutePathContext.Provider,
      { value: currentRoutePath },
      React.createElement(NodeRender, { node: node.node, ctx })
    );
  }

  return null;
}

export default RouteNodeRender;
