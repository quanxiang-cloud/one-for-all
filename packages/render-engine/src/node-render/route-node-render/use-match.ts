import { useEffect, useState } from 'react';
import { BehaviorSubject, distinctUntilChanged, map } from 'rxjs';
import type { Location } from 'history';

import { isParamHolder } from './utils';

export function exactlyCheck(path: string, currentRoutePath: string): boolean {
  const pathFragments = path.split('/');
  const routeFragments = currentRoutePath.split('/');
  if (pathFragments.length !== routeFragments.length) {
    return false;
  }

  return pathFragments.every((fragment, index) => {
    if (isParamHolder(routeFragments[index])) {
      return true;
    }

    return fragment === routeFragments[index];
  });
}

export function prefixCheck(path: string, currentRoutePath: string): boolean {
  const pathFragments = path.split('/');
  const routeFragments = currentRoutePath.split('/');
  if (pathFragments.length < routeFragments.length) {
    return false;
  }

  return routeFragments.every((fragment, index) => {
    if (isParamHolder(fragment)) {
      return true;
    }

    return fragment === pathFragments[index];
  });
}

function useMatch(location$: BehaviorSubject<Location>, currentRoutePath: string, exactly: boolean): boolean {
  const [match, setMatch] = useState(false);

  useEffect(() => {
    const subscribe = location$
      .pipe(
        map(({ pathname }): boolean => {
          return exactly ? exactlyCheck(pathname, currentRoutePath) : prefixCheck(pathname, currentRoutePath);
        }),
        distinctUntilChanged(),
      )
      .subscribe(setMatch);

    return () => subscribe.unsubscribe();
  }, []);

  return match;
}

export default useMatch;
