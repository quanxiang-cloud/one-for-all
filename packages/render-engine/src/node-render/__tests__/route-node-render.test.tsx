import React from 'react';
import { render } from '@testing-library/react';

import NodeRender from '..';
import initCTX from '../../ctx';
import { Props } from '../route-node-render';

test('route_node_render_item', async () => {
  const schema: SchemaSpec.Schema = {
    node: { id: 'some_node', type: 'html-element', name: 'div', props: {} },
    apiStateSpec: {},
    sharedStatesSpec: {},
  };
  const { ctx } = await initCTX({
    schema,
    apiStateSpec: {},
    sharedStatesSpec: {},
  });
  
  // to mock latest route state when url change;
  ctx.routeState = {
    action: 'push',
    location: {
      pathname: '/aaa/bbb/test1', // default pathname is '/'
      hash: '',
      search: '',
      state: '',
      key: 'default'
    }
  };

  const props: Props = {
    ctx,
    node: {
      id: 'test',
      type: 'route-node',
      path: '/aaa/bbb/test1',
      node: {
        id: 'el',
        type: 'html-element',
        name: 'i',
      }
    },
  };

  const { container } = render(<NodeRender {...props} />);
  expect(container.childNodes).toBeTruthy();
});

test('route_node_render_null', async () => {
  const schema: SchemaSpec.Schema = {
    node: { id: 'some_node', type: 'html-element', name: 'div', props: {} },
    apiStateSpec: {},
    sharedStatesSpec: {},
  };
  const { ctx } = await initCTX({
    schema,
    apiStateSpec: {},
    sharedStatesSpec: {},
  });
  
  // to mock latest route state when url change;
  ctx.routeState = {
    action: 'push',
    location: {
      pathname: '/aaa/bbb', // default pathname is '/'
      hash: '',
      search: '',
      state: '',
      key: 'default'
    }
  };

  const props: Props = {
    ctx: ctx,
    node: {
      id: 'test',
      type: 'route-node',
      path: '/aaa/bbb/test2',
      node: {
        id: 'iii',
        type: 'html-element',
        name: 'i',
      },
    },
  };

  const { container } = render(<NodeRender {...props} />);
  expect(container.firstChild).toBeFalsy();
});

test('render_nest_route_node_match_element', async () => {
  const schema: SchemaSpec.Schema = {
    node: { id: 'some_node', type: 'html-element', name: 'div', props: {} },
    apiStateSpec: {},
    sharedStatesSpec: {},
  };
  const { ctx } = await initCTX({
    schema,
    apiStateSpec: {},
    sharedStatesSpec: {},
  });

  // to mock latest route state when url change;
  ctx.routeState = {
    action: 'push',
    location: {
      pathname: '/aaa/bbb/test2/newtttt/sadnewtttt', // default pathname is '/'
      hash: '',
      search: '',
      state: '',
      key: 'default'
    }
  };

  const props: Props = {
    ctx: ctx,
    node: {
      id: 'A',
      type: 'route-node',
      path: 'aaa/bbb/test2',
      node: {
        id: 'a',
        type: 'html-element',
        name: 'div',
      },
      children: [
          {
          id: 'AA',
          type: 'route-node',
          path: 'newt',
          node: {
            id: 'qqq',
            type: 'html-element',
            name: 'i',
          },
        },
        {
          id: 'AB',
          type: 'route-node',
          path: 'newtttt',
          node: {
            id: 'qqq1',
            type: 'html-element',
            name: 'ul',
          },
          children: [
            {
              id: 'ABC',
              type: 'route-node',
              path: 'sadnewtttt',
              node: {
                id: 'asdqqq1',
                type: 'html-element',
                name: 'li',
              },
            }
          ],
        }
      ]
    },
  };

  const { container } = render(<NodeRender {...props} />);
  expect(container).toMatchSnapshot();
});
