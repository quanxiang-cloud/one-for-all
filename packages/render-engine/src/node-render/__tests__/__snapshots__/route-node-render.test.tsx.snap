// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render_nest_route_node_match_element 1`] = `
<div>
  <div>
    <div>
      <span />
    </div>
  </div>
</div>
`;
